[libc++] Implement comparison operators for `tuple` added in C++23

frederick-vs-ja opened the pull request:
And constrain the new `operator==` since C++26.

This patch implements parts of P2165R4, P2944R3, and a possibly improved resolution of LWG3882. Currently, libstdc++ and MSVC STL constrain the new overloads in the same way.

Also set feature-test macro `__cpp_lib_constrained_equality` and add related release note, as P2944R3 will completed with this patch.

Fixes #136765. Fixes #136770. Fixes #105424.

frederick-vs-ja committed:
[libc++] Implement comparison operators for `tuple` added in C++23

And constrain the new `operator==` since C++26.

This patch implements parts of P2165R4, P2944R3, and a possibly improved
resolution of LWG3882. Currently, libstdc++ and MSVC STL constrain the
new overloads in the same way.

frederick-vs-ja committed:
[libc++] Set feature-test macro `__cpp_lib_constrained_equality`

And add a release note for P2944R3 and P3379R0.

frederick-vs-ja committed:
Potential workarounds for pre-CWG2369 Clang and module builds

frederick-vs-ja commented on libcxx/include/tuple:
```diff
@@ -220,7 +220,9 @@ template <class... Types>
 #  include <__config>
 #  include <__cstddef/size_t.h>
 #  include <__fwd/array.h>
+#  include <__fwd/complex.h>
 #  include <__fwd/pair.h>
+#  include <__fwd/subrange.h>
```
It seems that modules builds need these inclusions to make `std::get` visible.

ldionne replied:
How about including `<__fwd/get.h>` instead?

ldionne replied:
I agree that seems overkill, but IMO that's better than including forward declarations for seemingly unrelated classes like `complex` and `subrange`. What's the modules error you got?

frederick-vs-ja commented on libcxx/include/tuple:
```diff
@@ -288,6 +291,72 @@ _LIBCPP_BEGIN_NAMESPACE_STD
+template <class _Tp, class _Up, class _IndexSeq>
+struct __tuple_common_comparison_category_impl {};
+template <class _Tp, class _Up, size_t... _Indices>
+  requires(tuple_size_v<_Tp> == tuple_size_v<_Up>) && requires {
```
`(tuple_size_v<_Tp> == tuple_size_v<_Up>) &&` is added for CWG2369. The version of Clang used for CI hasn't got #122423 landed.

h-vetinari replied:
This could be a `FIXME:` comment?

ldionne commented on libcxx/docs/Status/Cxx23Papers.csv:
```diff
@@ -60,7 +60,7 @@
 "`P1899R3 <https://wg21.link/P1899R3>`__","``stride_view``","2022-07 (Virtual)","","",""
 "`P2093R14 <https://wg21.link/P2093R14>`__","Formatted output","2022-07 (Virtual)","|Complete|","18",""
-"`P2165R4 <https://wg21.link/P2165R4>`__","Compatibility between ``tuple``, ``pair`` and ``tuple-like`` objects","2022-07 (Virtual)","|Partial|","","Only the part for ``zip_view`` is implemented."
+"`P2165R4 <https://wg21.link/P2165R4>`__","Compatibility between ``tuple``, ``pair`` and ``tuple-like`` objects","2022-07 (Virtual)","|Partial|","","Changes of ``tuple``, ``adjacent_view``, and ``cartesian_product_view`` are not yet implemented."
```
In principle, this would need to be an update to the `BEGIN-RST-NOTES` tag on https://github.com/llvm/llvm-project/issues/105200. But I agree this is not ergonomic since you really want the change to be associated to this PR and blameable somehow. Perhaps this means that we should make the status pages be the canonical representation of our conformance state, or perhaps it means that the version of the notes in the status pages should be the canonical ones? As I said in #148874, I'd love to find a workflow that works and is relatively simple.

ldionne commented on libcxx/test/std/utilities/tuple/tuple.tuple/tuple.rel/three_way.pass.cpp:
```diff
@@ -28,6 +30,15 @@ TEST_MSVC_DIAGNOSTIC_IGNORED(4242 4244)
 #include <tuple>
 #include <type_traits> // std::is_constant_evaluated
 
+#if TEST_STD_VER >= 23
```
IMO it's fine to always unconditionally include all these headers.

frederick-vs-ja replied:
Changed.

ldionne commented on libcxx/include/tuple:
```diff
@@ -288,6 +291,72 @@ _LIBCPP_BEGIN_NAMESPACE_STD
+template <size_t _Ip>
+struct __tuple_equal {
+  template <class _Tp, class _Up>
+  _LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR_SINCE_CXX14 bool operator()(const _Tp& __x, const _Up& __y) {
```
I know this is pre-existing, but isn't there a more compile-time-efficient way of implementing this? Fold expressions for example?

ldionne reviewed (commented):
Thanks for the patch!

frederick-vs-ja committed:
Merge branch 'main' into p2165r4-comparison

frederick-vs-ja committed:
Tweak implementation of `operator==`

Use `if constexpr` and `__all`.

frederick-vs-ja committed:
Rename `_Indices` to simpler `_Is`

frederick-vs-ja committed:
Add missed synopsis comments

frederick-vs-ja committed:
Unconditionally include headers in `tuple.rel/three_way.pass.cpp`

ldionne commented on libcxx/include/tuple:
```diff
@@ -288,6 +296,66 @@ _LIBCPP_BEGIN_NAMESPACE_STD
+}
+
+#    if _LIBCPP_STD_VER >= 26
+template <class _Tp, class _Up, class _IndexSeq>
+inline constexpr bool __can_tuple_compare_equal_impl = false;
```
I think you can simplify this a bit by doing this instead:

```c++
template <class _Tp, class _Up, class _IndexSeq = make_index_sequence<tuple_size_v<_Tp>>>
inline constexpr bool __can_tuple_compare_equal = false;


template <class _Tp, class _Up, size_t... _Is>
   requires(tuple_size_v<_Tp> == tuple_size_v<_Up>)
inline constexpr bool __can_tuple_compare_equal<_Tp, _Up, index_sequence<_Is...>> =
  __all<requires(const tuple_element_t<_Is, _Tp>& __t, const tuple_element_t<_Is, _Up>& __u) {
    { __t == __u } -> __boolean_testable;
  }...>::value;
```

I think this lets you get rid of the `_impl` layer.

frederick-vs-ja committed:
Apply @ldionne's suggestions

- include `<__fwd/get.h>` to get all `std::get` overloads
- eliminate `__can_tuple_compare_equal_impl` layer

frederick-vs-ja committed:
Merge branch 'main' into p2165r4-comparison

frederick-vs-ja committed:
Merge branch 'main' into p2165r4-comparison

frederick-vs-ja committed:
`FIXME` comments for CWG2369

ldionne reviewed (approved):
LGTM.

I would still like to backport this, it simplifies our support status for several papers that were otherwise kind of spotty.

ldionne committed:
Apply suggestions from code review

ldionne commented:
Merging since the only change was inside two comments.

ldionne commented:
/cherry-pick 4a509f853fa4821ecdb0f6bc3b90ddd48794cc8c

philnik777 commented:
This patch increases the pre-compiled size of `<tuple>` by 25%. It would be great if you could take a look at that.
